sympy/solvers/tests/test_solveset.py

```diff
@@ -2323,10 +2229,123 @@ def test_solve_logarithm():
     y = Symbol('y', positive=True)
     assert _solve_logarithm(log(x)*log(y), 0, x, S.Reals) == FiniteSet(1)
 
-# end of logarithmic tests
+#  end of logarithmic tests
+
+
+# lambert tests for solveset
+def test_solve_lambert():
+    yr = Symbol('yr', real=True)
+    b = Symbol('b')
+    a = Symbol('a')
+
+    assert solveset_real(x*exp(x) - 1, x) == FiniteSet(LambertW(1))
+    assert solveset_real(exp(x) + x, x) == FiniteSet(-LambertW(1))
+    assert solveset_real(x + 2**x, x) == \
+        FiniteSet(-LambertW(log(2))/log(2))
+
+    assert solveset_real(3*x + log(4*x), x) == \
+        FiniteSet(LambertW(Rational(3, 4))/3)
+
+    assert solveset_real(a/x + exp(x/2), x) == Complement(Intersection(\
+        FiniteSet(2*LambertW(-a/2)), S.Reals), FiniteSet(0))
```

Comment:
If `a = 1/2` then there would be two solutions for `x`.

Reply:
Yes, but I think it could be done in other PR's also. ( like more detailed specifications about the intervals for symbols along with complex infrastructure).

Reply:
I think that this needs to be fixed in this PR because otherwise this returns incorrect results.

Reply:
> If a = 1/2 then there would be two solutions for x.


>>> a = 1/2
>>> solveset(a/x + exp(x/2), x,Reals)
FiniteSet(-0.714805912362778, 2*LambertW(-0.25, -1))
>>> a = Rational(1, 2)
>>> solveset(a/x + exp(x/2), x,Reals)
FiniteSet(2*LambertW(-1/4), 2*LambertW(-1/4, -1))

Reply:
> IMHO, this should not be disturbed. If we try to give 2 solutions for symbols then all of them should be updated accordingly!

This is why I originally said that we should not use code from `solve` in `solveset`. The `solve` code can return incomplete solutions but in solveset it is very important that a complete solution is always returned. If the user doesn't care about complete solutions then they can use `solve`. The whole point of `solveset` is that the output is correct and complete even if given in a format that is difficult to actually use (e.g. an `Intersection`).

If it is possible to improve both `solve` and `solveset` here by making the Lambert code more complete then that is good. If the current Lambert code returns incomplete solutions then we can not use it directly in `solveset`.